k2/csrc/cuda/debug.h

```diff
@@ -0,0 +1,382 @@
+ * @endcode
+ */
+#ifndef K2_STATIC_ASSERT
+#define K2_STATIC_ASSERT(exp, msg) static_assert(exp, msg)
+#endif
+
+/**
+ * @brief Check if the expression is true.
+ *
+ * @details Implemented by `assert`, which is supported by host and device.
+ *
+ * @param[in] exp the boolean expression that should be true
+ *
+ * @note
+ *  `assert(exp)` in device, if exp == 0, the kernel excution is halted.
+ *  If the program is run within a debugger, this triggers a breakpoint and the
+ *  debugger can be used to inspect the current state of the device.
+ *  Otherwise, each thread for which expression is equal to zero prints a
+ *  message to stderr after synchronization with the host via
+ *  `cudaDeviceSynchronize(),cudaStreamSynchronize(),cudaEventSynchronize()`.
+ *  The format of this message is as follows:
+ *
+ *  @code
+ *  <filename>:<line number>:<function>:
+ *  block: [blockId.x,blockId.x,blockIdx.z],
+ *  thread: [threadIdx.x,threadIdx.y,threadIdx.z]
+ *  Assertion `<expression>` failed.
+ *  @endcode
+ *
+ *  @code{.cpp}
+ *  K2_ASSERT(1 == 1);
+ *  @endcode
+ */
+#ifndef K2_ASSERT
+#define K2_ASSERT(exp) assert(exp)
+#endif
+
+/**
+ * @brief Check if two arguments are equal.
+ *
+ * @details Implemented by `assert`, which is supported by host and device.
+ *
+ * @param[in] a left argument to compare
+ * @param[in] b right argument to compare
+ *
+ * @note
+ *  Assertions are for debugging purposes. They can affect performance
+ *  and it is therefore recommended to disable them in production code.
+ *  `assert` can be disabled at compile time by defining the `NDEBUG`
+ *  preprocessor macro before including assert.h.
+ *  Thus, `K2_MAKE_ERROR_CHECK` is ignored here.
+ *
+ *  @code{.cpp}
+ *  K2_CHECK_EQ(1, 1);
+ *  @endcode
+ */
+#ifndef K2_CHECK_EQ
+#define K2_CHECK_EQ(a, b) assert( a == b )
+#endif
+
+/**
+ * @fn
+ *  __host__ __device__ __forceinline__ cudaError_t
+ *  _K2CudaDebug(cudaError_t error,
+ *               const char *filename,
+ *               int line)
+ *
+ *
+ * @brief This is a error checking function, with context information.
+ *        It's not designed to called by users, but inner macros.
+ *
+ * @details
+ *  If K2_MAKE_ERROR_CHECK is defined and error is not cudaSuccess,
+ *  the corresponding error message is printed to:
+ *    - host: stderr
+ *    - device: stdout in device
+ *  along with the supplied source context.
+ *  It's used to make other macros convenient.
+ *
+ * @param[in] error         an enum type indicating CUDA errors.
+ * @param[in] filename      the source filename that the error comes from.
+ * @param[in] line          the code line that the error happened.
+ * @return                  Pass the input CUDA error.
+ *
+ * @code{.cpp}
+ * _K2CudaDebug(cudaGetLastError(), __FILE__, __LINE__);
+ * @endcode
+ */
+__host__ __device__ __forceinline__ cudaError_t _K2CudaDebug(
+    cudaError_t error,
+    const char *filename,
+    int line) {
+#ifdef K2_MAKE_ERROR_CHECK
+  (void)filename;
```

Comment:
Why do you put these two lines here?

Reply:
They should suppress the warning: variables not used. But I made a mistake here. These two lines should be out of the above macro. 
Fixed

Reply:
They are used in `printf` or `fprintf`; why will the compiler complain?

Reply:
Well, as it was that situation, but during our latest comments, I removed the K2_MAKE_ERROR_CHECK as Dan required.